## Week 10 and 11 Assignment - DATASCI200 Introduction to Data Science Programming, UC Berkeley MIDS

Write code in this Jupyter Notebook to solve the following problems. Please upload this **Notebook** with your solutions to your GitHub repository and provide a link in the last question in gradescope. 

Assignment due date: 11:59PM PT the night before the Week 12 Live Session. Do **NOT** push/upload the data file. 

## Objectives

- Explore and glean insights from a real dataset using pandas
- Practice using pandas for exploratory analysis, information gathering, and discovery
- Practice cleaning data and answering questions

## General Guidelines:

- This is a **real** dataset and so it may contain errors and other pecularities to work through
- This dataset is ~218mb, which will take some time to load (and probably won't load in Google Sheets or Excel)
- If you make assumptions, annotate them in your responses
- While there is one code/markdown cell positioned after each question as a placeholder, some of your code/responses may require multiple cells
- Double-click the markdown cells that say for example **1a answer here:** to enter your written answers. If you need more cells for your written answers, make them markdown cells (rather than code cells)
- This homework assignment is not autograded because of the variety of responses one could give. 
  - Please upload this notebook to the autograder page and the TAs will manually grade it. 
  - Ensure that each cell is run and outputs your answer for ease of grading! 
  - Highly suggest to do a `restart & run all` before uploading your code to ensure everything runs and outputs correctly.
  - Answers without code (or code that runs) will be given 0 points.
- **This is meant to simulate real world data so you will have to do some external research to determine what some of the answers are!** 

## Dataset

You are to analyze campaign contributions to the 2016 U.S. presidential primary races made in California. Use the csv file located here: https://drive.google.com/file/d/1Lgg-PwXQ6TQLDowd6XyBxZw5g1NGWPjB/view?usp=sharing. You should download and save this file in the same folder as this notebook is stored.  This file originally came from the U.S. Federal Election Commission (https://www.fec.gov/).

**DO NOT PUSH THIS FILE TO YOUR GITHUB REPO!**

- Best practice is to not have DATA files in your code repo. As shown below, the default load is outside of the folder this notebook is in. If you change the folder where the file is stored please update the first cell!
- If you do accidentally push the file to your github repo - follow the directions here to fix it: https://docs.google.com/document/d/15Irgb5V5G7pKPWgAerH7FPMpKeQRunbNflaW-hR2hTA/edit?usp=sharing

Documentation for this data can be found here: https://drive.google.com/file/d/11o_SByceenv0NgNMstM-dxC1jL7I9fHL/view?usp=sharing

## Data Questions

You are working for a California state-wide election campaign. Your boss wants you to examine historic 2016 election contribution data to see what zipcodes are more supportive of fundraising for your candidate. 

Your boss asks you to filter out some of the records:
- Only use primary 2016 contribution data (more like how your race is).
- Concentrate on Bernie Sanders as a candidate (most a like your candidate)

The questions your boss wants answered is:
- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
- What day(s) of the month do most people donate?

## Setup

Run the cell below as it will load the data into a pandas dataframe named `contrib`. Note that a custom date parser is defined to speed up loading. If Python were to guess the date format, it would take even longer to load.

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

# These commands below set some options for pandas and to have matplotlib show the charts in the notebook
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format

# Define a date parser to pass to read_csv
d = lambda x: datetime.strptime(x, '%d-%b-%y')

# Load the data
# We have this defaulted to the folder OUTSIDE of your repo - please change it as needed
contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)

# Note - for now, it is okay to ignore the warning about mixed types. 

/opt/conda/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3433: DtypeWarning: Columns (6,11,12) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


***
## 1. Initial Data Checks (50 points)

First we will take a preliminary look at the data to check that it was loaded correctly and contains the info we need.

The questions to answer at the end of this section:
- Do we have the correct # of columns and rows. 
- Do the records contain data for the questions we want to answer 
- What columns are important? 
- What columns can be dropped?
- What are the data problems?

**1a.** Print the *shape* of the data. Does this match the expectation? (2 points)

In [2]:
# 1a YOUR CODE HERE
contrib.shape

(1125659, 18)

- **1a answer here:**  This is in line with expectations since this is probably a survey. Having one million subjects is a massive sample size. 18 indicates the number of variables we are looking at.

**1b.** Print a list of column names. Are all the columns included that are in the documentation? (2 points)

In [3]:
# 1b YOUR CODE HERE
contrib.columns

Index(['cmte_id', 'cand_id', 'cand_nm', 'contbr_nm', 'contbr_city',
       'contbr_st', 'contbr_zip', 'contbr_employer', 'contbr_occupation',
       'contb_receipt_amt', 'contb_receipt_dt', 'receipt_desc', 'memo_cd',
       'memo_text', 'form_tp', 'file_num', 'tran_id', 'election_tp'],
      dtype='object')

- **1b answer here:**  Yup. All the columns included are in the documentation.

**1c** Print out the first five rows of the dataset. How do the columns `cand_id`, `cand_nm` and `contbr_st` look? (3 points)

In [4]:
# 1c YOUR CODE HERE
contrib.head(5)

cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st     contbr_zip            contbr_employer  \
0       LARKSPUR        CA 949,391,913.00                        NaN   
1        CAMBRIA        CA 934,284,638.00                        NaN   
2        FONTANA        CA 923,371,507.00                        NaN   
3      CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26          NaN       X   
1            RETIRED             200.00       2016-04-20          NaN       X   
2            RETIRED               5.00       2016-04-02          NaN       X   
3  SOFTWARE ENGINEER              40.00       2016-03-04          NaN     NaN   
4         PHARMACIST              35.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
0               * HILLARY VICTORY FUND    SA18   1091718     C4768722   
1               * HILLARY VICTORY FUND    SA18   1091718     C4747242   
2               * HILLARY VICTORY FUND    SA18   1091718     C4666603   
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   

  election_tp  
0       P2016  
1       P2016  
2       P2016  
3       P2016  
4       P2016

- **1c answer here:** It looks like there are duplicates for cand_id and cand_nm. The contribution state so far have been in california only.

**1d.** Print out the values for the column `election_tp`. In your own words, based on the documentation, what information does the `election_tp` variable contain? Do the values in the column match the documentation? (3 points)

In [5]:
# 1d YOUR CODE HERE
contrib['election_tp']

0          P2016
1          P2016
2          P2016
3          P2016
4          P2016
           ...  
1125654    P2016
1125655    P2016
1125656    P2016
1125657    P2016
1125658    P2016
Name: election_tp, Length: 1125659, dtype: object

- **1d answer here:** It shows what type of election this takes place, which is the 2016 presidential election. It is in line with the documentation since it looks at the election type.

**1e.** Print out the datatypes for all of the columns. What are the datatypes for the `contbr_zip`, `contb_receipt_amt`, `contb_receipt_dt`? (5 points)

In [6]:
# 1e YOUR CODE HERE
contrib.dtypes

cmte_id                      object
cand_id                      object
cand_nm                      object
contbr_nm                    object
contbr_city                  object
contbr_st                    object
contbr_zip                   object
contbr_employer              object
contbr_occupation            object
contb_receipt_amt           float64
contb_receipt_dt     datetime64[ns]
receipt_desc                 object
memo_cd                      object
memo_text                    object
form_tp                      object
file_num                      int64
tran_id                      object
election_tp                  object
dtype: object

- **1e answer here:** datatypes for contbr_zip, contb_receipt_amt, contb_receipt_dt are string object, float, and datetime respectively.

**1f.** What columns have the most non-nulls?  Would you recommend to drop any columns based on the number of nulls? (5 points)

In [7]:
# 1f YOUR CODE HERE
contrib.isna().sum()

cmte_id                    0
cand_id                    0
cand_nm                    0
contbr_nm                  0
contbr_city               26
contbr_st                  0
contbr_zip                95
contbr_employer       157902
contbr_occupation      10399
contb_receipt_amt          0
contb_receipt_dt           0
receipt_desc         1110614
memo_cd               981391
memo_text             624511
form_tp                    0
file_num                   0
tran_id                    0
election_tp             1425
dtype: int64

- **1f answer here:** cmte_id, cand_id, cand_nm, contbr_nm, contbr_st, contb_recept_amt, contb_receipt_dt, form_tp, file_num, and train_id have no null values. Yes. I would drop receipt_desc, memo_cd, and probably memo_text since more than half of them have null values. 

**1g.** A column we know that we want to use is the cand_nm column.  From the documentation each candidate is a unique candidate id also. Check data quality of `cand_id` column to see if it matches `cand_nm` column. Specifically check to ensure our targetted candidate 'Bernard Sanders' always has the same cand_id throughout. Any issues with `cand_nm` matching `cand_id`? (5 points)

In [8]:
# 1g YOUR CODE HERE
sanders = contrib[contrib['cand_nm'].str.contains("Sanders")]
sanders['cand_id'].value_counts()

P60007168    407172
Name: cand_id, dtype: int64

- **1g answer here:** We checked that Bernie Sanders only has one candidate id throughout all observations. 

**1h.** Another area to check is to make sure all of the records are from California. Check the `contbr_st` column - are there any records outside of California based on `contbr_st`? (5 points)

In [9]:
# 1h YOUR CODE HERE
contrib['contbr_st'].value_counts()

CA    1125659
Name: contbr_st, dtype: int64

- **1h answer here:** no. there are only CA

**1i.** The next column to check for the analysis is the `tran_id` column. This column could be the primary key so look for duplicates. How many duplicate entries are there? Any pattern for why are there duplicate entries? (5 points)

In [10]:
# 1i YOUR CODE HERE
np.sum((contrib['tran_id'].value_counts() > 1).values)

3419

- **1i answer here:** There are duplicate entries.

**1j.** Another column to check is the `contb_receipt_amt` that shows the donation amounts. How many negative donations are included? What do negative donations mean? Please show at least pull a few rows to look at the records with negative donations. Do these records match with the expectation of why a negative donation would happen? (5 points)

In [11]:
# 1j YOUR CODE HERE
((contrib['contb_receipt_amt'] < 0).values).sum()
contrib[contrib['contb_receipt_amt'] < 0].head(5)

cmte_id    cand_id                    cand_nm           contbr_nm  \
19   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
23   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
81   C00574624  P60006111  Cruz, Rafael Edward 'Ted'    JOLLIFF, RICHARD   
190  C00574624  P60006111  Cruz, Rafael Edward 'Ted'    NOWELL, DIANA L.   
213  C00574624  P60006111  Cruz, Rafael Edward 'Ted'  LICHTY, ANDREW MR.   

                contbr_city contbr_st     contbr_zip  \
19                    CHICO        CA 959,289,507.00   
23                    CHICO        CA 959,289,507.00   
81                    CHICO        CA 959,289,507.00   
190  RANCHO SANTA MARGARITA        CA 926,884,928.00   
213               SAN DIEGO        CA 921,096,720.00   

                        contbr_employer     contbr_occupation  \
19                        SELF EMPLOYED               RANCHER   
23                        SELF EMPLOYED               RANCHER   
81                        SELF EMPLOYED               RANCHER   
190  CAPISTRAND UNIFIED SCHOOL DISTRICT  LIBRARIAN TECHNICIAN   
213                       SELF EMPLOYED           REAL ESTATE   

     contb_receipt_amt contb_receipt_dt              receipt_desc memo_cd  \
19              -25.00       2016-04-29  REDESIGNATION TO GENERAL       X   
23             -150.00       2016-04-29  REDESIGNATION TO GENERAL       X   
81              -60.00       2016-04-14  REDESIGNATION TO GENERAL       X   
190            -100.00       2016-04-11  REDESIGNATION TO GENERAL       X   
213             -25.00       2016-04-30  REDESIGNATION TO GENERAL       X   

                    memo_text form_tp  file_num         tran_id election_tp  
19   REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1826482B       P2016  
23   REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1826483B       P2016  
81   REDESIGNATION TO GENERAL   SA17A   1077664   SA17A.1827494       P2016  
190  REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1639830B       P2016  
213  REDESIGNATION TO GENERAL   SA17A   1077664  SA17A.1826888B       P2016

- **1j answer here:** There are 11,896 negative donations. It has to do with reallocation or have some value under receipt_desc and memo text. This indicates some redistribution of funds.

**1k.** One more column to look at is the date of donation column. Are there any dates outside of the primary period (defined as 1 Jan 2014 to 7 June 2016)? Are the dates well-formatted for our analysis? (5 points)

In [12]:
# 1k YOUR CODE HERE
contrib[contrib['contb_receipt_dt'] < datetime(2014, 1, 1)]

cmte_id    cand_id       cand_nm          contbr_nm contbr_city  \
9932  C00458844  P60006723  Rubio, Marco  WHEELER, MARY MS.    ATHERTON   
9994  C00458844  P60006723  Rubio, Marco  WHEELER, MARY MS.    ATHERTON   

     contbr_st     contbr_zip contbr_employer  contbr_occupation  \
9932        CA 940,273,415.00   SELF-EMPLOYED  INTERIOR DESIGNER   
9994        CA 940,273,415.00   SELF-EMPLOYED  INTERIOR DESIGNER   

      contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
9932             -20.00       2013-11-05          NaN       X   
9994              20.00       2013-11-05          NaN       X   

                        memo_text form_tp  file_num             tran_id  \
9932  TRANSFER FROM RUBIO VICTORY    SA18   1029436  SA18.631526.2.0615   
9994  TRANSFER FROM RUBIO VICTORY    SA18   1029436  SA18.631526.3.0615   

     election_tp  
9932       P2016  
9994       G2016

In [13]:
contrib[contrib['contb_receipt_dt'] > datetime(2016, 6, 7)]

cmte_id    cand_id                    cand_nm           contbr_nm  \
14673    C00574624  P60006111  Cruz, Rafael Edward 'Ted'     RANDALL, DICK J   
14682    C00605568  P20002671              Johnson, Gary           BOAL, ROB   
14697    C00605568  P20002671              Johnson, Gary         LEE, JASCHA   
14698    C00605568  P20002671              Johnson, Gary         LEE, JASCHA   
14700    C00575795  P00003392    Clinton, Hillary Rodham         BAKER, JUDY   
...            ...        ...                        ...                 ...   
1123992  C00575795  P00003392    Clinton, Hillary Rodham  RANDALL, NANCY ENA   
1123993  C00575795  P00003392    Clinton, Hillary Rodham       WRIGHT, LORNA   
1123994  C00575795  P00003392    Clinton, Hillary Rodham      BALES, REBECCA   
1123995  C00575795  P00003392    Clinton, Hillary Rodham   BELLINGER, MARTHA   
1123996  C00580100  P80001571           Trump, Donald J.  DELLA DONNA, KADEE   

         contbr_city contbr_st     contbr_zip                 contbr_employer  \
14673      CUPERTINO        CA      95,014.00                             NaN   
14682       OAK PARK        CA      91,377.00                         VENDAVO   
14697     SANTA CRUZ        CA      95,065.00                           YAHOO   
14698     SANTA CRUZ        CA      95,065.00                           YAHOO   
14700      VACAVILLE        CA 956,873,433.00                             NaN   
...              ...       ...            ...                             ...   
1123992  BAKERSFIELD        CA      933124485                         RETIRED   
1123993      BENICIA        CA      945101743                             NaN   
1123994     PASADENA        CA      911051700  PASADENA HEALTHCARE CONSULTING   
1123995    CLAREMONT        CA      917112219                             NaN   
1123996   LONG BEACH        CA          90803                   SELF-EMPLOYED   

        contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc  \
14673                 NaN          -5,400.00       2016-06-30       Refund   
14682            ENGINEER              30.00       2016-07-20          NaN   
14697             S/W ENG              88.45       2016-07-14          NaN   
14698             S/W ENG             100.00       2016-07-24          NaN   
14700             RETIRED               5.00       2016-07-25          NaN   
...                   ...                ...              ...          ...   
1123992  REGISTERED NURSE               5.00       2016-08-02          NaN   
1123993           RETIRED              50.00       2016-08-09          NaN   
1123994      HOSPITAL CFO             100.00       2016-08-31          NaN   
1123995           RETIRED              75.00       2016-08-31          NaN   
1123996       REAL ESTATE             250.00       2016-07-06          NaN   

        memo_cd               memo_text form_tp  file_num        tran_id  \
14673       NaN                     NaN   SB28A   1096256    SB28A.90932   
14682       NaN                     NaN   SA17A   1096305    SA17A.50793   
14697       NaN                     NaN   SA17A   1096305    SA17A.51414   
14698       NaN                     NaN   SA17A   1096305    SA17A.51415   
14700         X  * HILLARY VICTORY FUND    SA18   1109498       C8807328   
...         ...                     ...     ...       ...            ...   
1123992     NaN                     NaN   SA17A   1100718       C8235857   
1123993     NaN                     NaN   SA17A   1100718       C8997947   
1123994     NaN                     NaN   SA17A   1100718       C9829807   
1123995     NaN                     NaN   SA17A   1100718       C9862797   
1123996     NaN                     NaN   SA17A   1104813  SA17A.1627816   

        election_tp  
14673           NaN  
14682         G2016  
14697         G2016  
14698         G2016  
14700         P2016  
...             ...  
1123992       G2016  
1123993       G2016  
1123994       G2016  
1123995

- **1k answer here:** They are already in datetime. So this is well formatted for our analysis. There are only 2 datapoints before January 1st 2014 but over 400,000 after June 7th 2016.

**1l.** Finally, answer the initial questions in the cells below (5 points)

**1l.1** Do we have the correct # of columns and rows.

- **1l.1 answer here:** Yes. we have the correct number of columns and rows.

**1l.2** Do the records contain data for the questions we want to answer?

- **1l.2 answer here:** Yes, our records contain the necessary data for the questions on zipcode and dates we want to answer.

**1l.3** What columns are important?

- **1l.3 answer here:** zipcode, dates, contribution receipt amount, candidate number or id since we are only looking at bernie sanders.

**1l.4** What columns can be dropped?

- **1l.4 answer here:** everything else 

**1l.5** What are the data problems?

- **1l.5 answer here:** The only data problem is negative donations. They should fall under another category since these cash flow represents another form of transaction outside of donations.

**1l.6** List any assumptions so far:

- **1l.6 answer here:** One assumption I have made is that after ensuring our targetted candidate, Bernie Sanders, have consistent candidate ID, i did not check for other candidates but only assumed the other candidates are fine if Bernie Sanders, one of the more controversial candidates, is free from data errors.

***
## 2. Data filtering and data quality fixes (30 points)

Now that we have a basic understanding of the data, let's filter out the records we don't need and fix the data.

**2a.** From the dataset filter out (remove) any election_tp not in the primary election. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [2]:
# 2a YOUR CODE HERE
p2016_contrib = contrib[contrib['election_tp'] == 'P2016']
p2016_contrib.head(5)

cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st     contbr_zip            contbr_employer  \
0       LARKSPUR        CA 949,391,913.00                        NaN   
1        CAMBRIA        CA 934,284,638.00                        NaN   
2        FONTANA        CA 923,371,507.00                        NaN   
3      CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26          NaN       X   
1            RETIRED             200.00       2016-04-20          NaN       X   
2            RETIRED               5.00       2016-04-02          NaN       X   
3  SOFTWARE ENGINEER              40.00       2016-03-04          NaN     NaN   
4         PHARMACIST              35.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
0               * HILLARY VICTORY FUND    SA18   1091718     C4768722   
1               * HILLARY VICTORY FUND    SA18   1091718     C4747242   
2               * HILLARY VICTORY FUND    SA18   1091718     C4666603   
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   

  election_tp  
0       P2016  
1       P2016  
2       P2016  
3       P2016  
4       P2016

**2b.** From the dataset filter out (remove) any candidate that is not Bernie Sanders. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [3]:
# 2b YOUR CODE HERE
sanders_contrib = p2016_contrib[p2016_contrib['cand_nm'] == 'Sanders, Bernard']
sanders_contrib.head(5)

cmte_id    cand_id           cand_nm         contbr_nm    contbr_city  \
3  C00577130  P60007168  Sanders, Bernard         LEE, ALAN      CAMARILLO   
4  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
5  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
6  C00577130  P60007168  Sanders, Bernard    LEOPARD, PATTI          VISTA   
8  C00577130  P60007168  Sanders, Bernard      LEPKE, KELLY    WESTMINSTER   

  contbr_st     contbr_zip            contbr_employer  contbr_occupation  \
3        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS  SOFTWARE ENGINEER   
4        CA 902,784,310.00   VERICOR ENTERPRISES INC.         PHARMACIST   
5        CA 902,784,310.00   VERICOR ENTERPRISES INC.         PHARMACIST   
6        CA 920,842,849.00  ONSITE ENERGY CORPORATION    PROJECT MANAGER   
8        CA 926,833,846.00                       NONE       NOT EMPLOYED   

   contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
3              40.00       2016-03-04          NaN     NaN   
4              35.00       2016-03-05          NaN     NaN   
5             100.00       2016-03-06          NaN     NaN   
6              25.00       2016-03-04          NaN     NaN   
8              10.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   
5  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKYBXV4   
6  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKW04C1   
8  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3H59   

  election_tp  
3       P2016  
4       P2016  
5       P2016  
6       P2016  
8       P2016

**2c.** The `contbr_zip` column is not formatted well for our analysis. Make a new zipcode column that is the five-digit zipcodes. Filter out any records outside of California based on the zipcode. Print/show the shape of the dataframe after the filtering is complete. (10 points).

- You will have to research what the valid 5-digit zipcodes for California are!

In [4]:
# 2c YOUR CODE HERE
zipcode = sanders_contrib
zipcode['zip'] = zipcode['contbr_zip'].astype(int)
zipcode['ca_zipcode'] = zipcode['zip'].astype(str).str[:5].astype(int)
ca_zip = zipcode[zipcode['ca_zipcode'] >= 89061]
ca_zip = ca_zip[ca_zip['ca_zipcode'] <= 97635]
ca_zip.head(5)

/tmp/ipykernel_284/3514574897.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  zipcode['zip'] = zipcode['contbr_zip'].astype(int)
/tmp/ipykernel_284/3514574897.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  zipcode['ca_zipcode'] = zipcode['zip'].astype(str).str[:5].astype(int)


cmte_id    cand_id           cand_nm         contbr_nm    contbr_city  \
3  C00577130  P60007168  Sanders, Bernard         LEE, ALAN      CAMARILLO   
4  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
5  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
6  C00577130  P60007168  Sanders, Bernard    LEOPARD, PATTI          VISTA   
8  C00577130  P60007168  Sanders, Bernard      LEPKE, KELLY    WESTMINSTER   

  contbr_st     contbr_zip            contbr_employer  contbr_occupation  \
3        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS  SOFTWARE ENGINEER   
4        CA 902,784,310.00   VERICOR ENTERPRISES INC.         PHARMACIST   
5        CA 902,784,310.00   VERICOR ENTERPRISES INC.         PHARMACIST   
6        CA 920,842,849.00  ONSITE ENERGY CORPORATION    PROJECT MANAGER   
8        CA 926,833,846.00                       NONE       NOT EMPLOYED   

   contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
3              40.00       2016-03-04          NaN     NaN   
4              35.00       2016-03-05          NaN     NaN   
5             100.00       2016-03-06          NaN     NaN   
6              25.00       2016-03-04          NaN     NaN   
8              10.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   
5  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKYBXV4   
6  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKW04C1   
8  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3H59   

  election_tp        zip  ca_zipcode  
3       P2016  930111214       93011  
4       P2016  902784310       90278  
5       P2016  902784310       90278  
6       P2016  920842849       92084  
8       P2016  926833846       92683

**2d.** The receipt amount column has negative donations. After talking with your team, a decision was made that the best course of action is to remove these negative values so that the donation count and amount is more accurate. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [5]:
# 2d YOUR CODE HERE
positive = ca_zip[ca_zip['contb_receipt_amt'] >= 0]
positive.head(5)

cmte_id    cand_id           cand_nm         contbr_nm    contbr_city  \
3  C00577130  P60007168  Sanders, Bernard         LEE, ALAN      CAMARILLO   
4  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
5  C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE  REDONDO BEACH   
6  C00577130  P60007168  Sanders, Bernard    LEOPARD, PATTI          VISTA   
8  C00577130  P60007168  Sanders, Bernard      LEPKE, KELLY    WESTMINSTER   

  contbr_st     contbr_zip            contbr_employer  contbr_occupation  \
3        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS  SOFTWARE ENGINEER   
4        CA 902,784,310.00   VERICOR ENTERPRISES INC.         PHARMACIST   
5        CA 902,784,310.00   VERICOR ENTERPRISES INC.         PHARMACIST   
6        CA 920,842,849.00  ONSITE ENERGY CORPORATION    PROJECT MANAGER   
8        CA 926,833,846.00                       NONE       NOT EMPLOYED   

   contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
3              40.00       2016-03-04          NaN     NaN   
4              35.00       2016-03-05          NaN     NaN   
5             100.00       2016-03-06          NaN     NaN   
6              25.00       2016-03-04          NaN     NaN   
8              10.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   
5  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKYBXV4   
6  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKW04C1   
8  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3H59   

  election_tp        zip  ca_zipcode  
3       P2016  930111214       93011  
4       P2016  902784310       90278  
5       P2016  902784310       90278  
6       P2016  920842849       92084  
8       P2016  926833846       92683

**2e.** From the dataset drop any columns that won't be used in the analysis. Print/show the shape of the dataframe after the dropping is complete. What columns did you drop and why? (5 points)

In [6]:
# 2e YOUR CODE HERE
final = positive[['cand_id', 'cand_nm', 'contb_receipt_dt', 'contb_receipt_amt', 'ca_zipcode']]
final.head(5)

cand_id           cand_nm contb_receipt_dt  contb_receipt_amt  ca_zipcode
3  P60007168  Sanders, Bernard       2016-03-04              40.00       93011
4  P60007168  Sanders, Bernard       2016-03-05              35.00       90278
5  P60007168  Sanders, Bernard       2016-03-06             100.00       90278
6  P60007168  Sanders, Bernard       2016-03-04              25.00       92084
8  P60007168  Sanders, Bernard       2016-03-05              10.00       92683

- **2e answer here:** I dropped everything else other than candidate id, candidate name, contribution date, contribution amount, and zipcode. Everything else were irrelevant for our study.

**2f.** List any assumptions that you made up to this point:

- **2f answer here:** After some cleaning and filtering, all assumptions from initial exploration are no longer needed.

***
## 3. Answering the questions (20 points)

Now that the data is cleaned and filterd - let's answer the two questions from your boss!

**3a.** Which zipcode had the highest count of contributions and the most dollar amount? (10 points)

In [7]:
# 3a YOUR CODE HERE
final.groupby(['ca_zipcode'])['contb_receipt_amt'].agg(['sum','count']).sort_values(by='count', ascending=False)

sum  count
ca_zipcode                  
94110      294,061.13   4046
94114      208,602.37   3043
94117      201,545.43   2884
95060      145,414.63   2849
94611      150,232.12   2516
...               ...    ...
93530          300.00      1
93590          100.00      1
92616           27.00      1
92079           28.00      1
93190           50.00      1

[1834 rows x 2 columns]

- **3a answer here:**  The zipcode with 94110 had the greatest dollar amount for contributions at 294,061.13 as well as the greatest number of contributions at 4046.

**3b.** What day(s) of the month do most people donate? (10 points)

In [8]:
# 3b YOUR CODE HERE
final.groupby(['contb_receipt_dt'])['contb_receipt_amt'].count().sort_values(ascending=False)

contb_receipt_dt
2016-02-29    8458
2016-03-31    8438
2016-05-31    7084
2016-04-30    6690
2016-03-09    6495
2016-03-14    6397
2016-03-30    6390
2016-04-08    5611
2016-03-16    4898
2016-04-06    4824
2016-04-15    4524
2016-03-29    4404
2016-02-10    4313
2016-04-20    3940
2016-05-04    3912
2016-06-08    3754
2016-03-13    3623
2016-04-17    3622
2016-05-03    3610
2016-03-06    3585
2016-04-13    3551
2016-03-23    3462
2016-04-09    3460
2016-06-30    3407
2016-04-14    3361
2016-03-27    3294
2016-03-17    3241
2016-05-10    3198
2016-05-18    3177
2016-02-02    3130
2016-03-28    3076
2016-02-23    3003
2016-01-29    2975
2016-02-09    2862
2016-04-05    2860
2016-03-12    2778
2016-03-08    2756
2016-04-18    2744
2016-02-26    2718
2016-04-01    2708
2016-02-28    2609
2016-04-29    2607
2016-05-11    2585
2016-04-26    2582
2016-04-27    2579
2016-05-22    2565
2016-03-05    2544
2016-03-21    2542
2016-04-19    2447
2016-03-25    2392
2016-04-16    2369
2016-02-20    

- **3b answer here:** Most people donate near important election events. The day that received the most number of donations was 29th of February.

## If you have feedback for this homework, please submit it using the link below:

http://goo.gl/forms/74yCiQTf6k